Make qcv2 target&leader observer execute in parallel

congqixia opened the pull request:
- Add `taskDispatcher` to submit and run task async safely
- Change `LeaderObeserver` and `TargetObserver` schedule and manual check action to submitting task into dispatcher
- Fix logic problem in collection observer when manual check return false

See also #27494
/kind bug

congqixia commented:
/run-cpu-e2e

yah01 commented on internal/querycoordv2/observers/leader_observer.go:
```diff
@@ -57,28 +58,10 @@ func (o *LeaderObserver) Start() {
-			}
-		}
-	}()
+	go o.schedule(ctx)
```
The caller should maintain the waitgroup by itself, not the callee
```golang
go func() {
defer o.wg.Done()
o.schedule(ctx)
}()
```

congqixia replied:
nope, `wg` shall be `Add`ed and `Wait`ed by the owner, which is the observer, otherwize there will be race on it

congqixia replied:
fixed

yah01 commented on internal/querycoordv2/observers/task_dispatcher.go:
```diff
@@ -0,0 +1,101 @@
+func newTaskDispatcher[K comparable](runner task[K]) *taskDispatcher[K] {
+	return &taskDispatcher[K]{
+		tasks:      typeutil.NewConcurrentSet[K](),
+		pool:       conc.NewPool[any](8),
```
I believe 8 is too small for this, as the observer executes IO operations mostly, let's make it larger

congqixia replied:
no sure which value is suitable here

yah01 replied:
max(32, CpuNum() * 2)

congqixia replied:
changed to configurable item and increase default to 16

yah01 commented on internal/querycoordv2/observers/task_dispatcher.go:
```diff
@@ -0,0 +1,101 @@
+	d.cancel = cancel
+
+	d.wg.Add(1)
+	go d.schedule(ctx)
```
maintain the wg by caller

congqixia replied:
ditto

congqixia replied:
fixed

congqixia committed:
Make qcv2 target&leader observer execute in parallel

- Add `taskDispatcher` to submit and run task async safely
- Change `LeaderObeserver` and `TargetObserver` schedule and manual check action to submitting task into dispatcher
- Fix logic problem in collection observer when manual check return false

See also #27494

Signed-off-by: Congqi Xia <congqi.xia@zilliz.com>

yah01 reviewed (approved):
/lgtm
